trino-aws-proxy/src/test/java/io/trino/aws/proxy/server/testing/RequestRewriteUtil.java

```diff
@@ -0,0 +1,92 @@
+import io.trino.aws.proxy.server.testing.harness.BuilderFilter;
+import io.trino.aws.proxy.spi.credentials.Credential;
+import io.trino.aws.proxy.spi.credentials.Credentials;
+import software.amazon.awssdk.services.s3.S3Client;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+public final class RequestRewriteUtil
+{
+    public static final String TEST_CREDENTIAL_REDIRECT_BUCKET = "redirected-bucket-for-credential";
+    public static final String TEST_CREDENTIAL_REDIRECT_KEY = "redirected-key-for-credential";
+    public static final Credential CREDENTIAL_TO_REDIRECT = new Credential("credential-to-redirect", UUID.randomUUID().toString());
+
+    private RequestRewriteUtil() {}
+
+    public static class RequestRewriteSetupModule
+    {
+        @Inject
+        public RequestRewriteSetupModule(
+                TestingCredentialsRolesProvider credentialsRolesProvider,
+                @ForTesting Credentials testingCredentials,
+                @ForS3Container List<String> configuredBuckets,
+                @ForS3Container S3Client storageClient)
+        {
+            credentialsRolesProvider.addCredentials(Credentials.build(CREDENTIAL_TO_REDIRECT, testingCredentials.requiredRemoteCredential()));
+            configuredBuckets.forEach(bucket -> storageClient.createBucket(r -> r.bucket(getTargetName(bucket))));
+            storageClient.createBucket(r -> r.bucket(TEST_CREDENTIAL_REDIRECT_BUCKET));
+        }
+    }
+
+    public static class Filter
+            implements BuilderFilter
+    {
+        @Override
+        public TestingTrinoAwsProxyServer.Builder filter(TestingTrinoAwsProxyServer.Builder builder)
+        {
+            return builder
+                    .withS3RequestRewriter(Rewriter.class)
+                    .withProperty("aws.proxy.s3.path", "/api/some/s3/path")
+                    .addModule(binder -> binder.bind(RequestRewriteSetupModule.class).asEagerSingleton());
+        }
+    }
+
+    public static class Rewriter
+            implements TestingS3RequestRewriter
+    {
+        @Override
+        public S3RewriteResult testRewrite(Optional<Credentials> credentials, String bucketName, String keyName)
+        {
+            boolean redirectForTestCredential = credentials.map(requestCredentials -> requestCredentials.emulated().accessKey().equalsIgnoreCase(CREDENTIAL_TO_REDIRECT.accessKey())).orElse(false);
+            if (redirectForTestCredential) {
+                return new S3RewriteResult(TEST_CREDENTIAL_REDIRECT_BUCKET, keyName.isEmpty() ? "" : TEST_CREDENTIAL_REDIRECT_KEY);
+            }
+            return new S3RewriteResult(getTargetName(bucketName), getTargetName(keyName));
+        }
+    }
+
+    private static String getTargetName(String name)
+    {
+        return name.isEmpty() ? "" : "redirected-%s".formatted(name);
+    }
+
+    public static String getTargetBucket(TestingS3RequestRewriter requestRewriter, String sourceBucket, String sourceKey)
```

Comment:
nit: as we are injecting a custom class (`TestingS3RequestRewriter`) into the testing classes I think it makes sense for these to be member function